source/base/Promise.ooc

```diff
@@ -0,0 +1,92 @@
+import threading/Thread
+import os/Time
+
+PromiseState: enum {
+	Running
+	Completed
+	Cancelled
+	Error
+}
+
+Promise: abstract class <T> {
+	_result: T
+	_state: PromiseState
+	state ::= this _state
+
+	init: func
+
+	wait: abstract func -> T { this _result }
+
+	wait: virtual func ~timeout (seconds: Double) -> T { this _result }
+
+	cancel: virtual func -> Bool { false }
+
+	_execute: virtual func
+}
+
+ThreadedPromise : class <T> extends Promise <T> {
```

Comment:
Also, I think you misunderstood me last time - do write `Promise<T>` without spaces.